aea/cli/generate_all_protocols.py

```diff
@@ -0,0 +1,594 @@
+    result = subprocess.check_output(  # nosec
+        [str(shutil.which("git")), "diff", "--stat"]
+    )  # nosec
+    if len(result) > 0:
+        click.echo("Git working tree is dirty:")
+        click.echo(result.decode("utf-8"))
+        sys.exit(1)
+    else:
+        click.echo("All good!")
+
+
+def load_protocol_specification_from_string(
+    specification_content: str,
+) -> ProtocolSpecification:
+    """Load a protocol specification from string."""
+    file = StringIO(initial_value=specification_content)
+    config_loader = ConfigLoader(
+        "protocol-specification_schema.json", ProtocolSpecification
+    )
+    protocol_spec = config_loader.load_protocol_specification(file)
+    return protocol_spec
+
+
+def get_protocol_specification_from_readme(package_path: Path) -> str:
+    """Get the protocol specification from the package README."""
+    logger.info(f"Get protocol specification from README {package_path}")
+    readme = package_path / "README.md"
+    readme_content = readme.read_text()
+    enforce(
+        "## Specification" in readme_content,
+        f"Cannot find specification section in {package_path}",
+    )
+
+    search_result = SPECIFICATION_REGEX.search(readme_content)
+    enforce(
+        search_result is not None,
+        f"Cannot find specification section in README of {package_path}",
+    )
+    specification_content = cast(Match, search_result).group(0)
+    # just for validation of the parsed string
+    load_protocol_specification_from_string(specification_content)
+    return specification_content
+
+
+def subdirs(path: Path) -> Iterator[Path]:
+    """Get subdirectories of a path."""
+    return filter(methodcaller("is_dir"), path.iterdir())
+
+
+def find_protocols_in_local_registry(packages_dir: Path) -> Iterator[Path]:
+    """Find all protocols in local registry."""
+    authors = subdirs(packages_dir)
+    component_parents = chain(*map(subdirs, authors))
+    protocols_parent = filter(lambda p: p.name == PROTOCOLS_PLURALS, component_parents)
+    protocols = chain(*map(subdirs, protocols_parent))
+    return map(methodcaller("absolute"), protocols)
+
+
+def log(message: str, level: int = logging.INFO) -> None:
+    """Produce a logging message."""
+    logger.log(level, message)
+
+
+def log_separator(length: int = 100) -> None:
+    """Log separator."""
+    click.echo("=" * length)
+
+
+def run_cli(*args: Any, **kwargs: Any) -> None:
+    """Run a CLI command."""
+    log(f"Calling command {args} with kwargs {kwargs}")
+    return_code = subprocess.check_call(args, **kwargs)  # nosec
+    enforce(
+        return_code == 0,
+        f"Return code of {pprint.pformat(args)} is {return_code} != 0.",
+    )
+
+
+def run_aea(*args: Any, **kwargs: Any) -> None:
+    """
+    Run an AEA command.
+
+    :param args: the AEA command
+    :param kwargs: keyword arguments to subprocess function
+    """
+    run_cli(sys.executable, "-m", "aea.cli", *args, **kwargs)
+
+
+class AEAProject:
```

Comment:
This is a generic utility. I suspect we have something similar elsewhere in tests or aea.manager or aea.builder. If not, it should be extracted in high level component on aea level or sublevel so it can be easily reused.